apiserver/facades/client/resources/repository.go

```diff
@@ -140,13 +140,47 @@ func newCharmHubClient(client CharmHub, logger Logger) *charmHubClient {
 }
 
 func (ch *charmHubClient) ResourceInfo(curl *charm.URL, origin corecharm.Origin, name string, revision int) (charmresource.Resource, error) {
-	if origin.ID == "" {
-		return charmresource.Resource{}, errors.Errorf("empty charm ID")
+	var configs []charmhub.RefreshConfig
+	var err error
+
+	// Due to async charm downloading we may not always have a charm ID to
+	// use for getting resource info, however it is preferred. A charm name
```

Comment:
So the window here where there could be an issue using name is that someone renames the charm in the store before the async download has finished I think, right?
It seems implausible for the error condition, but could we add a small retry loop to wait for the ID to become available, ie allow a small, finite time for the async download to complete, and only fallback to name when absolutely necessary.

Reply:
This would be a complex change to go back to find the origin after the async charm download is complete. It'd be difficult to do this while not reversing the speed improvements provided by the async download as well.

I considered using the Name only, but given the future case of changing charm names, kept the ID as well. Will help with upgrades after a change in charm name.